### ISSUE TITLE
Adding temporary item

### ISSUE THREAD

**Issue by dvidic** [opensourcepos/opensourcepos#1815]
### Background information

__IMPORTANT: If you choose to ignore this issue report template, your issue will be closed as we cannot help without the requested information.__

Please make sure you tick (add an x between the square brackets with no spaces) the following check boxes:

- [] Reporting an issue of an unmodified OSPOS installation
- [] Checked [open and closed issues](https://github.com/opensourcepos/opensourcepos/issues?utf8=%E2%9C%93&q=is%3Aissue) and no similar issue was already reported (please make sure you searched!)
- [] Read [README](https://github.com/opensourcepos/opensourcepos/blob/master/README.md), [WHATS_NEW](https://github.com/opensourcepos/opensourcepos/blob/master/WHATS_NEW.txt) and [UPGRADE](https://github.com/opensourcepos/opensourcepos/blob/master/UPGRADE.txt)
- [] Read the [FAQ](https://github.com/opensourcepos/opensourcepos#faq) for any known install and/or upgrade gotchas (in specific PHP extensions installed)
- [] Read the [wiki](https://github.com/opensourcepos/opensourcepos/wiki)
- [] Executed any database upgrade scripts if an upgrade pre 3.0.0 (e.g. database/2.4_to_3.0.sql)
- [] Aware the installation code is in [bintray](https://bintray.com/jekkos/opensourcepos/opensourcepos/view/files?sort=updated&order=asc#files) (see README), and [GitHub master](https://github.com/opensourcepos/opensourcepos/tree/master) is for [developers only](https://github.com/opensourcepos/opensourcepos/wiki/Development-setup) and therefore not complete nor stable

### Installation information

- OSPOS version is: 3.2.0
- OSPOS git commit hash is: 
- PHP version is:     7.1
- MySQL or MariaDB version is:      MySQL 5.6,
- OS and version is:     Windows 10)
- WebServer is:    Apache 2.4

###New Feature

I'm suggesting new feature which is quite useful for quotes-> when making a quote add possibilty to add an item which is not on stock or was on the stock. That item would appear only on that particular quote and after quote creation it would be deleted. 
This could be done by adding a row on quote making for manual entering item data (code, name, cost, price, quantity)....
Is something like this possible?

**Comment by daN4cat** [comment]
I'm afraid I don't get why an item should be just temporary in a quote. What is the use case behind this story?

**Comment by dvidic** [comment]
The case is that you for example have a hardware store. And customer comes wants 50 items from catalog. You have 46 of them on stock in the warehouse, and 4 of them no. If you add new item it goes automatically on stock and appears on the item list. Also, sometimes customer needs some special combination of items which you can make but it is not an actual item, so its good to have possibilty to write any kind of name for an item in quote and it would dissapear after the quote is done/printed. Also, the process of adding new item is time consuming when customer needs quote for several item which are not on you stock. 


Also, is it possible to hide/show columns in quote module? For example item code or any other...?

I'm experiencing several real life features that are useful and not yet implemented in this great project, so please be patient with me ;)

**Comment by daN4cat** [comment]
@SteveIreland any view on this?

**Comment by daN4cat** [comment]
@dvidic one part is adding quantity, the other is adding a new item.
That's what I read from your description.

**Comment by dvidic** [comment]
@daN4cat You are reffering to temporary item? If that's the case, user should have possibilty to add custom name/quantity/code/prices in the field when making a qoute.

**Comment by daN4cat** [comment]
@dvidic I didn't implement that feature and neither I use it, so I don't have an opinion.

@SteveIreland implemented it and when he has the time he can answer.

This is the best I can say, but don't hold your breath to see this sorted / implemented soon.

**Comment by dvidic** [comment]
@daN4cat thnx for your answer, I hope @SteveIreland will give his oppinion on this or advice how to solve it by myself. 

Also, what about  hide/show columns in quote module? For example item code or any other...?

**Comment by dvidic** [comment]
@daN4cat , @SteveIreland , @odiea Guys, is there any progress on this? This feature is one of most valuable in quotes. Trust me, it's real life situation where this is very very useful...

**Comment by daN4cat** [comment]
As said, don't expect progress in this from my side. I don't use it and neither need it. Sorry...

**Comment by dvidic** [comment]
I see, but any help would be appreciated on this because this feature is needed on daily basis.

**Comment by daN4cat** [comment]
Appreciate but this is not a commercial product and we work on this when we find the time or it's useful for us in some context.
In my case for my wife's business. And she doesn't issue any invoice but just receipts.

**Comment by SteveIreland** [comment]
Sorry, I just noticed this request.  In our shop we simulated this feature by setting up a non-stock item to allow an alternate description.  We had a set of 10 items created just for this purpose (being that they could represent a multitude of items).   Then at sale time we would simply select one of the set and provide an alternate description and the price and other info at the point of sale.

Why would that not work for your situation?

**Comment by jekkos** [comment]
@SteveIreland exactly how I handle this case. Only my version does not have the non stock option yet. Just use a generic item and the chosen description will override the generic text.

**Comment by daN4cat** [comment]
Good idea...

**Comment by dvidic** [comment]
I see . But by this user cannot change item name/barcode "on the fly". Changing price, description is good, but item name/barcode are crucial...

Thank You guys for reply.

**Comment by dvidic** [comment]
Are there any movements/progress on this one?

**Comment by SteveIreland** [comment]
The @jekkos, @daN4cat and @odiea will need to first approve this, but I will commit to adding the ability override the item number and item name when the allow override flag is enabled.  It would require adding two additional fields to the sales_item table, and adding two additional input capable fields to the register (that would only appear if the item allows description override.

This is NOT a trivial change since it would also need to impact reports.  Additional impact analysis needs to be done and discussed before starting.

**Comment by odiea** [comment]
With an app config and you are up to it you have my vote to proceed.

**Comment by daN4cat** [comment]
No, sorry no more changes. I want to close 3.2.0. This is a custom case and I would not complicate OSPOS for 1 person case. They can branch and customise for their case. In the same way I did, I'm not adding my 3 custom fields and behaviour to the main code because it's custom to my needs.

**Comment by dvidic** [comment]
I'm sure this ain't one person case... Its a precious feature, I'm not sure why improvements are neglected. I'm using OSPOS and really want to make this as good as possible: I'm contributing as much as I can.

**Comment by SteveIreland** [comment]
I'd like to see 3.2 published too.  So after 3.2 is published maybe we can revisit this.  I can see the advantage of having something like this in a repair shop where a part might be fabricated on the fly and need to be reflected as a proper sale.  That would be "my" target use case for this feature.  It happens often.

**Comment by dvidic** [comment]
@SteveIreland Yes, that's the point! As I said before, I really like OSPOS and I'm trying to make it great, as much as I can. 
So this feature will be great improvement as it gives more freedom for users. Please. implement it....If I can help in some way, I'm here...
I looked at the way alternate description is inserted and tried to do the same for item number/name but I got stuck.

**Comment by dvidic** [comment]
@SteveIreland m8, have You had the chance to make something about this one?

tnx

**Comment by SteveIreland** [comment]
I am going to start this immediately after the release (in addition to resuming work on the purchasing system).

**Comment by dvidic** [comment]
@SteveIreland is it time for this one, please?

**Comment by SteveIreland** [comment]
I'm glad you brought it to my attention.  I've been buried in work related to my regular job and missed the release announcement party.  I'll celebrate tonight.  :wink:

I'll start preparing to work on this today.  My normal approach is to update the wiki with what I think I'm going to do and then fail to update it to reflect what I actually did.  So keep an eye on it.  Feedback is welcome.

**Comment by dvidic** [comment]
I'll contribute as much as I can. Thank You!

**Comment by dvidic** [comment]
@SteveIreland m8, is it time for this one now?

**Comment by SteveIreland** [comment]
It's in development.

**Comment by SteveIreland** [comment]
I'm close to finishing up the temporary item feature.   I'm including an overview in the next comment (which is also in the Wiki).

**Comment by SteveIreland** [comment]
3.3 Temporary Items
Temporary Items are created for the purpose selling an item that we do not want to manage in our "standard" collection of stock and non-stock items. They are created to record a sale of something that is not kept in stock and is not expected to be sold again. It could be that the item is expected to be sold again, but we are deferring setting it up properly until a later time.

Internally, a temporary item is maintained in the item file just like any other item. However, it has unique properties (other than an item type of "Temporary").

First, it is associated with a single order and is attached the order only when clicking on the New button from the sales register. Items that are added to the register can be deleted. If a temporary item is deleted from the register it is also removed from the items table.

Next, the temporary item id has a negative value for the item id. This is done to isolate temporary items from the rest of the set of items. This also provides a visual clue when looking at items that the item is a temporary item (in addition to the fact that the item_type has a value of '3' to indicate that it is a temporary item).

The approach taken to support temporary items is to insure the integrity of the reporting system which is dependent on joins from the sales_items table to the items table to retrieve item information. You can think of temporary items more as an extension of the sales_items table since that item only exists for the purpose of holding information about the temporary item for that sales_item.

**Comment by SteveIreland** [comment]
Pull request #1975 was submitted for this issue.

**Comment by SteveIreland** [comment]
I'm updating the ticket because there was a new requirement added to be able to convert a temporary item to a permanent item.  That requirement would have changed the approach used.  To explain, a primary requirement was to insure that temporary items didn't "co-mingle" with regular items.  To attain that requirement the approach used was to assign negative identifiers to temporary items.

In order to be able to make that happen the best way is to drop the negative item_id approach.  So I will be rewriting this.  I'm pulling the pull request back for now and will hopefully resubmit it by next weekend.

I'm not at all feeling bad about this last minute change because even though it sounded clever at the time, after I put it together my hope of being able to purge temporary items using the negative item id wasn't really fulfilled since temporary items can't be removed until the completed sales themselves are purged (and I don't think there is a purge for sales available in OSPOS yet).

I'll also be adding stronger validation to item type changes.  You should be able to convert a temporary item to a stock item or kit item, but you should never be able to change a stock item or kit item to a temporary item.

**Comment by SteveIreland** [comment]
Made progress on it over the last couple of days.  I'm just continuing to test it and refine it.  So I'll probably end up resubmitting the pull request prior to this coming weekend.

**Comment by SteveIreland** [comment]
I submitted a new pull request for this.  Hopefully this meets all requirements.

**Comment by jekkos** [comment]
I'd also prefer not touching db technical id's for realising business requirements. As you say think indeed it's better to use the item type field for this purpose, which will be more consistent with what already exists

**Comment by daN4cat** [comment]
@SteveIreland can you please test item_kits? Probably this temp item feature commit broke something. The view doesn't show anything and the error is:

```
A PHP Error was encountered
Severity: Warning

Message: A non-numeric value encountered

Filename: controllers/Item_kits.php

Line Number: 39

Backtrace:

File: /home/colibri2/public_html/pos3/application/controllers/Item_kits.php
Line: 39
Function: _error_handler

File: /home/colibri2/public_html/pos3/application/controllers/Item_kits.php
Line: 69
Function: _add_totals_to_item_kit

File: /home/colibri2/public_html/pos3/public/index.php
Line: 315
Function: require_once
```

**Comment by SteveIreland** [comment]
Will do, thanks.

**Comment by SteveIreland** [comment]
@daN4cat  Can you give me a clue as to what you were doing when you encountered this error?  I haven't been able to recreate it yet.

**Comment by SteveIreland** [comment]
@daN4cat This feels more like data corruption.  Temporary items shouldn't have anything to do with kit items, so I'm wondering if there was some finagling with item types that led to this issue.

**Comment by daN4cat** [comment]
Ok I'll check better if an issue from my side.

**Comment by daN4cat** [comment]
I'd probably create a new issue as that change was committed 2 months ago and it's in 3.2.{0, 1, 2, 3).

**Comment by daN4cat** [comment]
@SteveIreland did you manage to reproduce the issue?

**Comment by SteveIreland** [comment]
Not yet.  I was already heavily invested in the India GST prroject, so as soon as I get to the point where my sales functionality is mostly restored I was going to go upgrade my WampServer so that I can try out PHP 7.1.  The version I currently have installed tops out at 7.0.10

**Comment by SteveIreland** [comment]
Changed my mind and decided to go ahead and upgrade my Wampserver tonight and recreate my test environments.  I just hate being so far behind you.

**Comment by SteveIreland** [comment]
I tested a simple item kit scenario with PHP 7.1.16 (birthday package kit config) and did not encounter any issue.  I'll continue to test off and on over the coming week.

**Comment by daN4cat** [comment]
Ok I will try and debug over the weekend.

**Comment by daN4cat** [comment]
item_kits issue is fixed.

**Comment by daN4cat** [comment]
The temporary item is in master, shall we close this? Is it working as expected?
I didn't test myself, I'll check just for no regressions.

**Comment by SteveIreland** [comment]
You can close this. I’m going to retest over the weekend but if something crops up it will more likely be a different issue
